### PR TITLE
Refactor Language Server

### LINKED ISSUE #651: Refactor Language Server to use  CLI
### Issue To Be Solved
Currently Language Server executes scripts, sends transactions and deploys contracts on it's side without reporting back to editor extension. This creates little "holes" in developer experience and also prone to code duplication - since we can do exactly the same by calling Flow CLI with proper arguments and flags.

Additionally only arguments are supported via `pragma` docstrings, but not signers

### (Optional): Suggest A Solution
- Create code lenses to call action on editor side and pass arguments to it
- Allow to specify signers who will sign the transaction by name

### PR DESCRIPTION
**Pull request by MaxStalker**
Closes #651 
Depends on #656 

## Description

- Refactor codelenses to include signers and call client method
- Add emulator state into language server to allow updates to codelenses
______

For contributor use:

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Comment by turbolent** [comment]
@MaxStalker Nice work! 

Could you please open a separate PR for just the parser changes? The language server is a separate module depending on Cadence, so we need to update that first (and tag/release it)

File: languageserver/integration/codelenses.go [hunk 1]
```diff
@@ -83,23 +85,45 @@ func (i *FlowIntegration) showDeployContractAction(
...
+	}
+
+	for _, signers := range signersList {
+		title := fmt.Sprintf(
+			"%s %s %s %s %s",
+			prefixOK,
+			"Deploy contract",
+			name,
+			"to",
+			signers[0],
```
Comment (turbolent) [inline languageserver/integration/codelenses.go:111]: You can embed the constant parts directly in the format string:

```suggestion
			"%s Deploy contract %s to %s",
			prefixOK,
			name,
			signers[0],
```

File: languageserver/integration/codelenses.go [hunk 2]
```diff
@@ -108,36 +132,59 @@ func (i *FlowIntegration) showDeployContractAction(
...
-			Arguments: []interface{}{uri, name},
-		},
+	for _, signers := range signersList {
+		title := fmt.Sprintf(
+			"%s %s %s %s %s",
+			prefixOK,
+			"Deploy contract interface",
+			name,
+			"to",
+			signers[0],
```
Comment (turbolent) [inline languageserver/integration/codelenses.go:172]:
```suggestion
			"%s Deploy contract interface %s to %s",
			prefixOK,
			name,
			signers[0],
```

File: languageserver/integration/codelenses.go [hunk 3]
```diff
@@ -159,52 +206,64 @@ func (i *FlowIntegration) entryPointActions(
...
+	codelensRange := conversion.ASTToProtocolRange(position, position)
+	var codeLenses []*protocol.CodeLens
+
+	// If emulator is not up, we need to show single codelens proposing to start emulator
+	if i.emulatorState == EmulatorOffline {
+		title := fmt.Sprintf(
+			"%s %s",
+			prefixOffline,
+			"Emulator is Offline. Click here to start it",
+		)
```
Comment (turbolent) [inline languageserver/integration/codelenses.go:220]:
```suggestion
		title := fmt.Sprintf(
			"%s Emulator is Offline. Click here to start it",
			prefixOffline,
		)
```

File: languageserver/integration/codelenses.go [hunk 4]
```diff
@@ -159,52 +206,64 @@ func (i *FlowIntegration) entryPointActions(
...
 
-		title = fmt.Sprintf(
-			"submit with account 0x%s",
-			i.activeAddress.Hex(),
+	// If emulator is not up, we need to provide actionless codelens
+	if i.emulatorState == EmulatorStarting {
+		title := fmt.Sprintf(
+			"%s %s",
+			prefixStarting,
+			"Emulator is starting up. Please wait...",
```
Comment (turbolent) [inline languageserver/integration/codelenses.go:238]:
```suggestion
			"%s Emulator is starting up. Please wait \u2026",
			prefixStarting,
```

File: languageserver/integration/codelenses.go [hunk 5]
```diff
@@ -215,16 +274,64 @@ func (i *FlowIntegration) entryPointActions(
...
-			},
+		switch entryPointInfo.kind {
+
+		case entryPointKindScript:
+			if len(argumentList) > 0 {
+				title = fmt.Sprintf(
+					"%s %s %s",
+					prefixOK,
+					"Execute script with",
+					entryPointInfo.pragmaArgumentStrings[i],
```
Comment (turbolent) [inline languageserver/integration/codelenses.go:285]:
```suggestion
					"%s Execute script with %s",
					prefixOK,
					entryPointInfo.pragmaArgumentStrings[i],
```

File: languageserver/integration/codelenses.go [hunk 6]
```diff
@@ -215,16 +274,64 @@ func (i *FlowIntegration) entryPointActions(
...
+				)
+			} else {
+				title = fmt.Sprintf(
+					"%s %s",
+					prefixOK,
+					"Execute script",
```
Comment (turbolent) [inline languageserver/integration/codelenses.go:291]:
```suggestion
					"%s Execute script",
					prefixOK,
```

File: languageserver/integration/codelenses.go [hunk 7]
```diff
@@ -215,16 +274,64 @@ func (i *FlowIntegration) entryPointActions(
...
+		case entryPointKindTransaction:
+			for _, signers := range signersList {
+				if len(argumentList) > 0 {
+					title = fmt.Sprintf(
+						"%s %s %s %s %s",
+						prefixOK,
+						"Send with",
+						entryPointInfo.pragmaArgumentStrings[i],
+						"signed by ",
+						strings.Join(signers, " and "),
```
Comment (turbolent) [inline languageserver/integration/codelenses.go:314]:
```suggestion
						"%s Send with %s signed by %s",
						prefixOK,
						entryPointInfo.pragmaArgumentStrings[i],
						common.EnumerateWords(signers, "and"),
```

File: languageserver/integration/codelenses.go [hunk 8]
```diff
@@ -215,16 +274,64 @@ func (i *FlowIntegration) entryPointActions(
...
+					)
+				} else {
+					title = fmt.Sprintf(
+						"%s %s %s",
+						prefixOK,
+						"Send signed by",
+						strings.Join(signers, " and "),
```
Comment (turbolent) [inline languageserver/integration/codelenses.go:321]:
```suggestion
						"%s Send signed by %s",
						prefixOK,
						common.EnumerateWords(signers, "and"),
```

File: languageserver/integration/commands.go [hunk 9]
```diff
@@ -86,42 +97,36 @@ func (i *FlowIntegration) commands() []server.Command {
...
 		}
 
-		tx.AddRawArgument([]byte(transactionArgumentString))
+		transactionArguments = append(transactionArguments, strings.TrimSuffix(stringArgument, "\n"))
 	}
 
-	_, err = i.sendTransactionHelper(conn, i.activeAddress, tx)
+	// Pass arguments back to extension
+	codeArguments := fmt.Sprintf("%s", strings.Join(transactionArguments, ","))
+	signers := args[2].([]interface{})
```
Comment (turbolent) [inline languageserver/integration/commands.go:127]: Please validate the input and handle it gracefully:

```suggestion
	signers, ok := args[2].([]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid signers argument: %#+v", args[2])
	}
```

File: languageserver/integration/commands.go [hunk 10]
```diff
@@ -86,42 +97,36 @@ func (i *FlowIntegration) commands() []server.Command {
...
+	_, ok := args[1].([]interface{})
 	if !ok {
-		return nil, fmt.Errorf("could not find document for URI %s", uri)
+		return nil, fmt.Errorf("invalid transaction arguments: %#+v", args[1])
 	}
 
 	rawTransactionArguments, ok := args[1].([]interface{})
 	if !ok {
-		return nil, fmt.Errorf("invalid transaction arguments: %#+v", args[1])
+		return nil, fmt.Errorf("invalid script arguments: %#+v", args[1])
```
Comment (turbolent) [inline languageserver/integration/commands.go:112]: Why the change, this function is submitting a transaction, so the argument contains the transaction arguments, no?

File: languageserver/integration/commands.go [hunk 11]
```diff
@@ -86,42 +97,36 @@ func (i *FlowIntegration) commands() []server.Command {
...
-	if !ok {
-		return nil, fmt.Errorf("invalid URI argument: %#+v", args[0])
-	}
-
-	doc, ok := i.server.GetDocument(protocol.DocumentUri(uri))
+	_, ok := args[1].([]interface{})
 	if !ok {
-		return nil, fmt.Errorf("could not find document for URI %s", uri)
+		return nil, fmt.Errorf("invalid transaction arguments: %#+v", args[1])
 	}
```
Comment (turbolent) [inline languageserver/integration/commands.go:108]: This is now a duplicate of the check below, it can be removed here

File: languageserver/integration/commands.go [hunk 12]
```diff
@@ -137,82 +142,55 @@ func (i *FlowIntegration) executeScript(conn protocol.Conn, args ...interface{})
...
-		}
-		scriptArguments[i] = scriptArgument
+		scriptArguments = append(scriptArguments, strings.TrimSuffix(stringArgument, "\n"))
 	}
 
-	script := []byte(doc.Text)
-	res, err := i.flowClient.ExecuteScriptAtLatestBlock(context.Background(), script, scriptArguments)
-	if err != nil {
+	// Pass arguments back to extension
+	codeArguments := fmt.Sprintf("%s", strings.Join(scriptArguments, ","))
```
Comment (turbolent) [inline languageserver/integration/commands.go:167]: Why comma separate all the arguments and not just sent `scriptArguments` directly to the client?

File: languageserver/integration/commands.go [hunk 13]
```diff
@@ -86,42 +97,36 @@ func (i *FlowIntegration) commands() []server.Command {
...
 			return nil, fmt.Errorf("invalid transaction argument at index %d: %#+v", i, rawTransactionArgument)
 		}
 
-		tx.AddRawArgument([]byte(transactionArgumentString))
+		transactionArguments = append(transactionArguments, strings.TrimSuffix(stringArgument, "\n"))
 	}
 
-	_, err = i.sendTransactionHelper(conn, i.activeAddress, tx)
+	// Pass arguments back to extension
+	codeArguments := fmt.Sprintf("%s", strings.Join(transactionArguments, ","))
```
Comment (turbolent) [inline languageserver/integration/commands.go:126]: Why comma separate all the arguments and not just sent `scriptArguments` directly to the client?

File: languageserver/integration/commands.go [hunk 14]
```diff
@@ -137,82 +142,55 @@ func (i *FlowIntegration) executeScript(conn protocol.Conn, args ...interface{})
...
-					Message: fmt.Sprintf("Failed to execute script: %s", grpcErr.Message()),
-				})
-				return nil, nil
-			}
-		} else {
-			conn.LogMessage(&protocol.LogMessageParams{
-				Type:    protocol.Warning,
-				Message: fmt.Sprintf("Failed to execute script: %s", err.Error()),
-			})
-		}
```
Comment (turbolent) [inline languageserver/integration/commands.go:203]: 😍

File: languageserver/integration/commands.go [hunk 15]
```diff
@@ -137,82 +142,55 @@ func (i *FlowIntegration) executeScript(conn protocol.Conn, args ...interface{})
...
 		return nil, err
 	}
 
-	conn.LogMessage(&protocol.LogMessageParams{
-		Type:    protocol.Info,
-		Message: fmt.Sprintf("Executed script with result: %v", res),
-	})
+	emulatorState, ok := args[0].(float64)
+	if !ok {
+		return nil, errors.New("invalid argument")
```
Comment (turbolent) [inline languageserver/integration/commands.go:186]:
```suggestion
		return nil, fmt.Errorf("invalid emulator state argument: %#+v", args[0])
```

File: languageserver/integration/integration.go [hunk 16]
```diff
@@ -26,6 +26,12 @@ import (
 	"github.com/onflow/cadence/languageserver/server"
 )
 
+const (
+	EmulatorOffline = 1
+	EmulatorStarting = 2
+	EmulatorOnline = 3
+)
```
Comment (turbolent) [inline languageserver/integration/integration.go:33]:
```suggestion
type EmulatorState byte

const (
	EmulatorOffline EmulatorState = iota
	EmulatorStarting
	EmulatorOnline
)
```

File: languageserver/integration/integration.go [hunk 17]
```diff
@@ -34,6 +40,7 @@ type FlowIntegration struct {
 	activeAddress  flow.Address
 	serviceAddress flow.Address
 	entryPointInfo map[protocol.DocumentUri]entryPointInfo
+	emulatorState	byte
```
Comment (turbolent) [inline languageserver/integration/integration.go:43]:
```suggestion
	emulatorState	EmulatorState
```

File: languageserver/integration/initialization.go [hunk 18]
```diff
@@ -41,6 +41,9 @@ func (i *FlowIntegration) initialize(initializationOptions interface{}) error {
 		i.config.EmulatorAddr,
 		grpc.WithInsecure(),
 	)
+
+	i.emulatorState = byte(emulatorState)
```
Comment (turbolent) [inline languageserver/integration/initialization.go:45]:
```suggestion
	i.emulatorState = EmulatorState(emulatorState)
```

File: languageserver/integration/entrypoint.go [hunk 19]
```diff
@@ -107,5 +126,13 @@ func (i *FlowIntegration) updateEntryPointInfoIfNeeded(
 		parameters:            parameters,
 		pragmaArgumentStrings: pragmaArgumentStrings,
 		pragmaArguments:       pragmaArguments,
+		pragmaSignersStrings:  pragmaSigners,
 	}
 }
+
+func parseSignersList(signerList string) []string {
+	var re = regexp.MustCompile(`[a-zA-Z]+`)
```
Comment (turbolent) [inline languageserver/integration/entrypoint.go:134]: Please move this out of the function, so the regular expression is only compiled once and then re-used

File: languageserver/integration/config.go [hunk 20]
```diff
@@ -83,6 +83,11 @@ func configFromInitializationOptions(opts interface{}) (conf Config, err error)
 		return
 	}
 
+	emulatorState, ok = optsMap["emulatorState"].(float64)
+	if !ok {
+		emulatorState = 1
```
Comment (turbolent) [inline languageserver/integration/config.go:88]:
```suggestion
		emulatorState = EmulatorOffline
```

File: languageserver/integration/config.go [hunk 21]
```diff
@@ -47,30 +47,30 @@ type AccountPrivateKey struct {
 //
 // Returns an error if any fields are missing or malformed.
 //
-func configFromInitializationOptions(opts interface{}) (conf Config, err error) {
+func configFromInitializationOptions(opts interface{}) (conf Config, emulatorState float64, err error) {
 	optsMap, ok := opts.(map[string]interface{})
 	if !ok {
-		return Config{}, errors.New("invalid initialization options")
+		return Config{}, 1 , errors.New("invalid initialization options")
```
Comment (turbolent) [inline languageserver/integration/config.go:53]: You can make this line and the similar ones below more self-explaining by using the constants you have defined:

```suggestion
		return Config{}, EmulatorOffline, errors.New("invalid initialization options")
```

File: languageserver/integration/entrypoint.go [hunk 22]
```diff
@@ -33,6 +34,8 @@ const (
 	entryPointKindUnknown entryPointKind = iota
 	entryPointKindScript
 	entryPointKindTransaction
+	entryPointKindContract
+	entryPointKindContractInterface
```
Comment (turbolent) [inline languageserver/integration/entrypoint.go:38]: Contracts and contract interfaces are not entry points in the same sense as transactions and scripts. 

Maybe parse the signers for the contract (interface) in a new function instead of adding that to the entry point parsing function

File: languageserver/go.mod [hunk 23]
```diff
@@ -11,3 +11,7 @@ require (
 	github.com/stretchr/testify v1.7.0
 	google.golang.org/grpc v1.32.0
 )
+
+replace (
+	github.com/onflow/cadence => ../
+)
```
Comment (turbolent) [inline languageserver/go.mod:17]: Once the separate PR for the parser changes has been merged, we can update this PR and replace these lines with a version bump to the new Cadence version that includes the parser changes.

**Review by turbolent — COMMENTED** [review]
Nice work @MaxStalker 👏

File: languageserver/integration/commands.go (hunk 9 above)
Comment (MaxStalker) [reply languageserver/integration/commands.go:127]: I added proposed changes, though this code currently is not used, but will be refactored to use shared CLI library...

File: languageserver/integration/commands.go (hunk 10 above)
Comment (MaxStalker) [reply languageserver/integration/commands.go:112]: My bad, copy pasted portion of code from script execution and missed that later 🙇‍♂️

File: languageserver/integration/commands.go (hunk 12 above)
Comment (MaxStalker) [reply languageserver/integration/commands.go:167]: That's a really good question!
I think I was focused on getting a string on extension side, since I could not completely grasp the concept of how we communicate with it.

File: languageserver/integration/commands.go (hunk 13 above)
Comment (MaxStalker) [reply languageserver/integration/commands.go:126]: Same as with scripts 😅

File: languageserver/integration/initialization.go (hunk 18 above)
Comment (MaxStalker) [reply languageserver/integration/initialization.go:45]: That was exactly the thing I was looking for!

File: languageserver/integration/entrypoint.go (hunk 22 above)
Comment (MaxStalker) [reply languageserver/integration/entrypoint.go:38]: Got it. Will refactor it.

File: languageserver/integration/config.go [hunk 24]
```diff
@@ -47,30 +47,30 @@ type AccountPrivateKey struct {
...
 	optsMap, ok := opts.(map[string]interface{})
 	if !ok {
-		return Config{}, errors.New("invalid initialization options")
+		return Config{}, EmulatorOffline, errors.New("invalid initialization options")
 	}
 
 	emulatorAddr, ok := optsMap["emulatorAddress"].(string)
 	if !ok {
-		return Config{}, errors.New("initialization options: missing emulatorAddress field")
+		return Config{}, 1,  errors.New("initialization options: missing emulatorAddress field")
```
Comment (turbolent) [inline languageserver/integration/config.go:58]: Also use `EmulatorOffline` here and in the other return statements below

**Review by turbolent — COMMENTED** [review]
Nice cleanup 👍

**Comment by MaxStalker** [comment]
Closed in favor #751
